### PR TITLE
F/wep 87 social media metatags

### PR DESCRIPTION
**Pull request by 0815pascal**

### CODE REVIEW

**Review by penk85 — APPROVED** [review]

File: packages/api/src/graphql/article.ts [hunk 1]
```diff
@@ -88,6 +88,11 @@ export const GraphQLArticleInput = new GraphQLInputObjectType({
 
     hideAuthor: {type: GraphQLNonNull(GraphQLBoolean)},
 
+    socialMediaTitle: {type: GraphQLString},
+    socialMediaDescription: {type: GraphQLString},
+    socialMediaAuthorIDs: {type: GraphQLList(GraphQLID)},
```
Comment (roosnic1) [inline packages/api/src/graphql/article.ts:93]: should be same as `authorIDs`

File: packages/api/src/graphql/article.ts [hunk 2]
```diff
@@ -239,6 +266,32 @@ export const GraphQLPublicArticle: GraphQLObjectType<
     },
 
     breaking: {type: GraphQLNonNull(GraphQLBoolean)},
+
+    socialMediaTitle: {type: GraphQLString},
+    socialMediaDescription: {type: GraphQLString},
+    socialMediaAuthors: {
+      type: GraphQLList(GraphQLAuthor),
+      resolve: createProxyingResolver(({socialMediaAuthorIDs, hideAuthor}, args, {loaders}) => {
+        if (hideAuthor) {
```
Comment (roosnic1) [inline packages/api/src/graphql/article.ts:275]: return them regardless of `hideAuthor`

File: packages/editor/package.json [hunk 3]
```diff
@@ -142,6 +142,7 @@
         {
           "markupOnly": true,
           "ignoreAttribute": [
+            "eventKey",
```
Comment (roosnic1) [inline packages/editor/package.json:145]: i'm pretty sure you don't need this anymore.

File: packages/editor/src/client/panel/articleMetadataPanel.tsx [hunk 4]
```diff
@@ -49,17 +56,38 @@ export interface ArticleMetadataPanelProps {
...
+    socialMediaDescription,
+    socialMediaAuthors,
+    socialMediaImage
+  } = value
+
+  const [activeKey, setActiveKey] = useState(MetatagType.General)
 
   const [isChooseModalOpen, setChooseModalOpen] = useState(false)
   const [isEditModalOpen, setEditModalOpen] = useState(false)
+  const [chosenImageType, setChosenImageType] = useState('')
```
Comment (roosnic1) [inline packages/editor/src/client/panel/articleMetadataPanel.tsx:79]: why don't you set the image id here instead of which type.

File: packages/editor/src/client/routes/articleEditor.tsx [hunk 5]
```diff
@@ -141,7 +149,13 @@ export function ArticleEditor({id}: ArticleEditorProps) {
         breaking,
         authors: authors.filter(author => author != null) as AuthorRefFragment[],
         image: image || undefined,
-        hideAuthor
+        hideAuthor,
+        socialMediaTitle: socialMediaTitle ?? '',
```
Comment (roosnic1) [inline packages/editor/src/client/routes/articleEditor.tsx:153]: `''` is not the same as `undefined`

File: packages/editor/src/client/routes/pageEditor.tsx [hunk 6]
```diff
@@ -111,7 +125,10 @@ export function PageEditor({id}: PageEditorProps) {
           value: property.value,
           public: property.public
         })),
-        image: image || undefined
+        image: image || undefined,
+        socialMediaTitle: socialMediaTitle ?? '',
```
Comment (roosnic1) [inline packages/editor/src/client/routes/pageEditor.tsx:129]: `''` is not the same as `undefined`

**Review by roosnic1 — CHANGES_REQUESTED** [review]

File: examples/website/src/shared/route/articleTemplateContainer.tsx [hunk 7]
```diff
@@ -107,23 +120,32 @@ export function ArticleTemplateContainer({id, slug}: ArticleTemplateContainerPro
...
         {lead && <meta name="description" content={lead} />}
-
         <link rel="canonical" href={canonicalURL} />
-
-        <meta property="og:title" content={title} />
+        <meta property="og:title" content={socialMediaTitle ?? title} />
         <meta property="og:type" content="article" />
         <meta property="og:url" content={canonicalURL} />
-        {image && <meta property="og:image" content={image.ogURL} />}
+        <meta property="og:description" content={socialMediaDescription ?? ''} />
```
[the remark this answers is not in the document]
Comment (roosnic1) [reply examples/website/src/shared/route/articleTemplateContainer.tsx:127]: nvm

Comment (roosnic1) [inline examples/website/src/shared/route/articleTemplateContainer.tsx:127]: only render if `socialMediaDescription !== undefined`

File: packages/editor/src/client/blocks/types.ts [hunk 8]
```diff
@@ -151,6 +151,11 @@ export enum TeaserType {
   Page = 'page'
 }
 
+export enum MetatagType {
```
Comment (roosnic1) [inline packages/editor/src/client/blocks/types.ts:154]: rename to `MetaDataType`

File: packages/editor/src/client/panel/articleMetadataPanel.tsx (hunk 4 above)
Comment (roosnic1) [reply packages/editor/src/client/panel/articleMetadataPanel.tsx:79]: currentImage

File: packages/editor/package.json (hunk 3 above)
Comment (0815pascal) [reply packages/editor/package.json]: Done

File: packages/api-db-mongodb/src/migration.ts [hunk 9]
```diff
@@ -428,6 +428,109 @@ export const Migrations: Migration[] = [
         {$set: {'draft.hideAuthor': false}}
       )
     }
+  },
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply packages/api-db-mongodb/src/migration.ts:431]: Done

File: packages/api-db-mongodb/src/migration.ts [hunk 10]
```diff
@@ -428,6 +428,109 @@ export const Migrations: Migration[] = [
...
+        {
+          pending: {$ne: null},
+          'pending.socialMediaTitle': {$exists: false},
+          'pending.socialMediaDescription': {$exists: false},
+          'pending.socialMediaAuthors': {$exists: false},
+          'pending.socialMediaImage': {$exists: false}
+        },
+        {
+          $set: {
+            'pending.socialMediaTitle': false,
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply packages/api-db-mongodb/src/migration.ts]: Done

File: packages/editor/src/client/blocks/types.ts (hunk 8 above)
Comment (0815pascal) [reply packages/editor/src/client/blocks/types.ts:154]: Done

File: examples/website/src/shared/route/articleTemplateContainer.tsx [hunk 11]
```diff
@@ -107,23 +120,32 @@ export function ArticleTemplateContainer({id, slug}: ArticleTemplateContainerPro
...
+        {(image || socialMediaImage) && (
+          <meta property="og:image" content={socialMediaImage?.ogURL ?? image?.ogURL ?? ''} />
+        )}
+        {/* TODO: Add OpenGraph authors as soon as author profiles are implemented */}
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/route/articleTemplateContainer.tsx:131]: Done

File: packages/api/src/graphql/article.ts (hunk 1 above)
Comment (0815pascal) [reply packages/api/src/graphql/article.ts:93]: Done

File: packages/api/src/graphql/article.ts (hunk 2 above)
Comment (0815pascal) [reply packages/api/src/graphql/article.ts:275]: Done

File: examples/website/src/shared/route/articleTemplateContainer.tsx (hunk 7 above)
Comment (0815pascal) [reply examples/website/src/shared/route/articleTemplateContainer.tsx:127]: Done

File: examples/website/src/shared/route/articleTemplateContainer.tsx [hunk 12]
```diff
@@ -107,23 +120,32 @@ export function ArticleTemplateContainer({id, slug}: ArticleTemplateContainerPro
...
+        {/* <meta property="article:author" content="" /> */}
+
+        {socialMediaAuthors &&
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/route/articleTemplateContainer.tsx:134]: done

File: examples/website/src/shared/route/articleTemplateContainer.tsx [hunk 13]
```diff
@@ -232,7 +264,9 @@ export function PeerArticleTemplateContainer({
         <meta property="og:title" content={title} />
         <meta property="og:type" content="article" />
         <meta property="og:url" content={canonicalURL} />
-        {image && <meta property="og:image" content={image.ogURL} />}
+        {(image || socialMediaImage) && (
+          <meta property="og:image" content={socialMediaImage.ogURL ?? image.ogURL ?? ''} />
+        )}
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/route/articleTemplateContainer.tsx:269]: done

File: examples/website/src/shared/route/pageTemplateContainer.tsx [hunk 14]
```diff
@@ -109,10 +122,16 @@ export function PageTemplateContainer({slug, id}: PageTemplateContainerProps) {
         {description && <meta name="description" content={description} />}
         <link rel="canonical" href={canonicalURL} />
 
-        <meta property="og:title" content={title} />
+        <meta property="og:title" content={socialMediaTitle ?? title} />
+        {socialMediaDescription && (
+          <meta property="og:description" content={socialMediaDescription} />
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/route/pageTemplateContainer.tsx:127]: done

File: packages/editor/src/client/routes/pageEditor.tsx (hunk 6 above)
Comment (0815pascal) [reply packages/editor/src/client/routes/pageEditor.tsx:129]: Done

File: packages/editor/src/client/panel/articleMetadataPanel.tsx (hunk 4 above)
Comment (0815pascal) [reply packages/editor/src/client/panel/articleMetadataPanel.tsx]: done

File: packages/editor/src/client/routes/articleEditor.tsx (hunk 5 above)
Comment (0815pascal) [reply packages/editor/src/client/routes/articleEditor.tsx]: done

**Review by bishoymelek — APPROVED** [review]
Great work

File: packages/api/src/graphql/article.ts [hunk 15]
```diff
@@ -139,7 +139,7 @@ export const GraphQLArticleRevision = new GraphQLObjectType<ArticleRevision, Con
     socialMediaTitle: {type: GraphQLString},
     socialMediaDescription: {type: GraphQLString},
     socialMediaAuthors: {
-      type: GraphQLList(GraphQLAuthor),
+      type: GraphQLNonNull(GraphQLList(GraphQLAuthor)),
```
Comment (roosnic1) [inline packages/api/src/graphql/article.ts:142]: Missing one `GraphQLNonNull` between list and author. same as with properties or tags.

File: packages/api/src/graphql/article.ts [hunk 16]
```diff
@@ -268,7 +268,7 @@ export const GraphQLPublicArticle: GraphQLObjectType<
...
```
Comment (roosnic1) [inline packages/api/src/graphql/article.ts:271]: Missing one `GraphQLNonNull` between list and author. same as with properties or tags.

**Review by roosnic1 — CHANGES_REQUESTED** [review]
one last change

**Review by faebebin — DISMISSED** [review]

File: packages/api/src/graphql/article.ts (hunk 15 above)
Comment (0815pascal) [reply packages/api/src/graphql/article.ts]: Done

File: packages/api/src/graphql/article.ts (hunk 16 above)
Comment (0815pascal) [reply packages/api/src/graphql/article.ts]: Done

File: examples/website/src/shared/route/pageTemplateContainer.tsx [hunk 17]
```diff
@@ -122,9 +122,12 @@ export function PageTemplateContainer({slug, id}: PageTemplateContainerProps) {
         {description && <meta name="description" content={description} />}
         <link rel="canonical" href={canonicalURL} />
 
-        <meta property="og:title" content={socialMediaTitle ?? title} />
+        <meta property="og:title" content={socialMediaTitle !== '' ? socialMediaTitle : title} />
```
Comment (roosnic1) [inline examples/website/src/shared/route/pageTemplateContainer.tsx:125]: why did you change this?

File: examples/website/src/shared/route/pageTemplateContainer.tsx [hunk 18]
```diff
@@ -122,9 +122,12 @@ export function PageTemplateContainer({slug, id}: PageTemplateContainerProps) {
...
         {(socialMediaDescription || description) && (
-          <meta property="og:description" content={socialMediaDescription ?? description} />
+          <meta
+            property="og:description"
+            content={socialMediaDescription !== '' ? socialMediaDescription : description}
```
Comment (roosnic1) [inline examples/website/src/shared/route/pageTemplateContainer.tsx:129]: why did you change this?

**Review by roosnic1 — CHANGES_REQUESTED** [review]
As discussed

**Review by roosnic1 — APPROVED** [review]
👍🏾

File: packages/api-db-mongodb/src/migration.ts [hunk 19]
```diff
@@ -446,6 +446,45 @@ export const Migrations: Migration[] = [
...
     }
+  },
+  {
+    // Add social media metatags to article.
+    version: 8,
+    async migrate(db) {
+      await db.collection(CollectionName.Articles).updateMany(
+        {
+          pending: {$ne: null},
+          'pending.socialMediaAuthorsID': {$exists: false}
```
Comment (roosnic1) [inline packages/api-db-mongodb/src/migration.ts:457]: That field doesn't exist in the schema

File: packages/api-db-mongodb/src/migration.ts [hunk 20]
```diff
@@ -446,6 +446,45 @@ export const Migrations: Migration[] = [
...
+        {
+          $set: {
+            'pending.socialMediaAuthorsID': []
+          }
+        }
+      )
+      await db.collection(CollectionName.Articles).updateMany(
+        {
+          published: {$ne: null},
+          'published.socialMediaAuthorsID': {$exists: false}
```
Comment (roosnic1) [inline packages/api-db-mongodb/src/migration.ts:468]: That field doesn't exist in the schema

File: packages/api-db-mongodb/src/migration.ts [hunk 21]
```diff
@@ -446,6 +446,45 @@ export const Migrations: Migration[] = [
...
+        {
+          $set: {
+            'published.socialMediaAuthorsID': []
+          }
+        }
+      )
+      await db.collection(CollectionName.Articles).updateMany(
+        {
+          draft: {$ne: null},
+          'draft.socialMediaAuthorsID': {$exists: false}
```
Comment (roosnic1) [inline packages/api-db-mongodb/src/migration.ts:479]: That field doesn't exist in the schema

**Review by roosnic1 — CHANGES_REQUESTED** [review]
Bug in the migration

File: packages/api-db-mongodb/src/migration.ts (hunk 19 above)
Comment (0815pascal) [reply packages/api-db-mongodb/src/migration.ts]: Done

File: packages/api-db-mongodb/src/migration.ts (hunk 20 above)
Comment (0815pascal) [reply packages/api-db-mongodb/src/migration.ts]: Done

File: packages/api-db-mongodb/src/migration.ts (hunk 21 above)
Comment (0815pascal) [reply packages/api-db-mongodb/src/migration.ts]: Done

**Review by bishoymelek — APPROVED** [review]

**Review by roosnic1 — APPROVED** [review]
🎉
